code/headers/hardware_usart.hpp

```diff
@@ -1,175 +1,312 @@
+        /**
+         * @brief Write a uint8_t with the usart
+         *
+         * @param c data to send
+         * @return true if the data is sended correctly
+         * @return false if the data has not been sended correctly
+         */
         bool send(const uint8_t c) override {
-           send_byte(c);
-           return true;
-        }
+            while (!transmit_ready()) {
+                // wait until transmit is ready
+            }
 
-        /// @brief sends a char via usart
-        /// @param c: char to send
-        void putc(char c) override {
-            send_byte(c);
+            // write the byte in the Transmit holding register
+            detail::usart::port<Bus>->US_THR = c;
+
+            return true;
         }
 
-        /// @brief recieve byte bia usart
-        /// @return received byte
+        /**
+         * @brief Get a data from the usart
+         *
+         * @return uint8_t
+         */
         uint8_t receive() override {
             if (!input_buffer.size()) {
                 return 0;
             }
 
-            return input_buffer.copy_and_pop();
-        }
-
-        /// @brief check if char is available
-        /// @return true if char is available false if not
-        bool char_available() override {
-            return (available() > 0);
+            return input_buffer.copy_and_pop_front();
         }
 
-        /// @brief receive char via usart
-        /// @return char received
-        char getc() override {
-            if (char_available()) {
-                return receive();
-            }
-            return 0;
+        /**
+         * @brief returns how much data is available
+         *
+         * @return unsigned int amount of bytes
+         */
+        unsigned int available() override {
+            return input_buffer.size();
         }
 
-        /// @brief returns amount of available data in buffer
-        /// @return amount of uint8_t's in buffer
-        unsigned int available() override {
-           if ((hardware_usart->US_CSR & 1) != 0) {
+        /**
+         * @brief Interrupt handler for receiving data when the cpu is doing
+         * other stuff. The user shouln't call this function as it will be
+         * handled through the interrupt controller.
+         *
+         * @warning This adds the data into the ringbuffer and will overwrite
+         * old data if not read once in a while.
+         *
+         */
+        constexpr static void _isr_handler() {
+            // read the status register
+            uint32_t status = detail::usart::port<Bus>->US_CSR &
+                              detail::usart::port<Bus>->US_IMR;
+
+            // check if the rx ready bit is set
+            if ((status & US_CSR_RXRDY)) {
                 input_buffer.push(receive_byte());
             }
-
-            return input_buffer.size();
         }
-
     };
 }; // namespace r2d2::usart
 
+extern "C" {
+void __USART0_Handler() {
```

Comment:
`hwlib::cout` requires a uart too, do we need all the interrupt handlers? This will make the compiler generate a bunch of code, probably.

Reply:
This generates around 36 bytes of extra data in the text segment and ‭268 ‬bytes of data in the bss segment (RAM) after the first handler. I have not found a way to only include a handler when it is used. 

The uart will drop packets when nothing is polling the uart connection. As it will overwrite the old data when new data is read.

Reply:
Alright. I had a though about aliasing weak function definitions to a single function and maybe dispatching from there, but that won't apply here.